sdk/eventhub/Azure.Messaging.EventHubs/src/Authorization/EventHubSharedKeyCredential.cs

```diff
@@ -81,9 +109,19 @@ public sealed class EventHubSharedKeyCredential : TokenCredential
         ///
         /// <returns>A <see cref="SharedAccessSignatureCredential" /> based on the requested shared access key.</returns>
         ///
-        internal SharedAccessSignatureCredential ConvertToSharedAccessSignatureCredential(string eventHubResource,
-                                                                                          TimeSpan? signatureValidityDuration = default) =>
-            new SharedAccessSignatureCredential(new SharedAccessSignature(eventHubResource, SharedAccessKeyName, SharedAccessKey, signatureValidityDuration));
+        internal SharedAccessSignatureCredential AsSharedAccessSignatureCredential(string eventHubResource,
+                                                                                   TimeSpan? signatureValidityDuration = default)
+        {
+            if (SharedAccessSignatureCredential == null)
```

Comment:
I'm not sure that I understand why we wouldn't generate on demand.  Is there a benefit to caching the state?

Reply:
The main reason for caching is to allow ```UpdateSharedAccessKey``` to hold a reference to an existing credential so to call ```UpdateSharedAccessKey``` on it.

Without holding a reference, the caller may not use ```EventHubSharedKeyCredential``` in order to call the internal method ```SharedAccessSignatureCredential::UpdateSharedAccessKey```

Reply:
My apologies, I deleted the previous message I wrote as I believe you are referring to the ```else``` branch below. 

You have got a really good point. Great thinking.

I was thinking that if the method ```AsSharedAccessSignatureCredential``` was called in another scenario from the one we are already using, it would open the door to a bug?

When doing, I was thinking to this sequence of events:

A.  ```AsSharedAccessSignatureCredential``` is called, a new instance of ```SharedAccessSignatureCredential``` is returned and cached let's say in ```AmqpClient```
B. ```AsSharedAccessSignatureCredential``` is called again in the same EventHub instance
C. The user calls ```UpdateSharedAccessKey``` and the key would only be partially rotated in the client

That said, you noticed I am opening the door to the possibility of a user passing the same ```TokenCredential``` to different event hubs clients.

I will remove that code. Again, great thinking.

Reply:
If we weren't caching within the callers, I think this level would have made sense.   But, as you mentioned, it's the responsibility of those using the tokens to cache as needed and each request to `GetAccessToken` is an explicit request for "please get me a new value".   My thinking is just to be sure we honor that pattern.

Reply:
Would you be opposed to keeping the original name as well?   Since we're generating a new credential on demand,  I find the `AsXXX` misleading.  We're not changing the representation, we're essentially creating a new object based on the current state of this one.

Reply:
What about ```ToSharedAccessSignatureCredential```?

Reply:
Works for me.   Did a quick scan of the code, and it doesn't look like we do these conversions frequently enough that there's any established pattern.... so we're good if you want to go in that direction.

Reply:
> Not that I don't appreciate the kind words, but you're too quick with the praise for this one. 😄 There's a ton of merit in your scenario and I overlooked some of the shared cases. I think you've got them well covered in this form, unless I'm overlooking a corner case again.

Thank you, I was genuinely impressed by your catch. Nice sense of humour too.

> so we're good if you want to go in that direction

I think we are, nice points. Flagging the last thought I had.

Was looking at [```AsEnumerable```](
https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/System.Core/System/Linq/Enumerable.cs#L937
) in search for inspiration and it seems to always return the same reference.

So was thinking to flag ```AsNewSharedAccessSignatureCredential``` as a possibility?

Reply:
Looking things over, I don't know that `New` is accurate or necessary here.   You're correctly caching the value and returning the same instance so that we maintain a link and can update the values.   Unless I'm not seeing a scenario, I believe the only time that a new instance is created is on the "we don't already have one" case.   The internals for the credential's  `SharedAccessSignature` change, but that's an implementation detail that I don't think we need to reflect here.

Am I overlooking something or over thinking?